Source/ACE.Database/ShardDatabase.cs

```diff
@@ -461,68 +569,252 @@ private bool SaveObjectInternal(DatabaseTransaction transaction, AceObject aceOb
-            SaveAceObjectPropertiesAttribute2nd(transaction, aceObject.AceObjectId,
-                aceObject.AceObjectPropertiesAttributes2nd.Select(x => x.Value.GetVital(aceObject.AceObjectId)).ToList());
-            SaveAceObjectPropertiesSkill(transaction, aceObject.AceObjectId,
-                aceObject.AceObjectPropertiesSkills.Select(x => x.Value.GetAceObjectSkill(aceObject.AceObjectId)).ToList());
 
             return true;
         }
 
         private bool SaveAceObjectBase(DatabaseTransaction transaction, AceObject obj)
         {
-            transaction.AddPreparedInsertStatement<ShardPreparedStatement, AceObject>(ShardPreparedStatement.SaveAceObject, obj);
+            if (obj.IsDirty)
+            {
+                if (!obj.HasEverBeenSavedToDatabase)
+                    transaction.AddPreparedInsertStatement<ShardPreparedStatement, AceObject>(ShardPreparedStatement.SaveAceObject, obj);
+                else
+                    transaction.AddPreparedUpdateStatement<ShardPreparedStatement, AceObject>(ShardPreparedStatement.UpdateAceObject, obj);
+            }
+
             return true;
         }
 
         // FIXME(ddevec): These are a lot of functions that essentially do the same thing... but the SharedPreparedStatement.--- makes them a pain to tempalte/reduce
         private bool SaveAceObjectPropertiesInt(DatabaseTransaction transaction, uint aceObjectId, List<AceObjectPropertiesInt> properties)
         {
-            transaction.AddPreparedInsertListStatement<ShardPreparedStatement, AceObjectPropertiesInt>(ShardPreparedStatement.InsertAceObjectPropertiesInt, properties);
+            // calling ToList forces the enumerable evaluation so that we can call .Remove from within the loop
+            var theDirtyOnes = properties.Where(p => p.IsDirty).ToList();
+
+            foreach (var prop in theDirtyOnes)
+            {
+                if (prop.HasEverBeenSavedToDatabase)
+                {
+                    if (prop.PropertyValue == null)
+                    {
+                        // delete it
+                        transaction.AddPreparedDeleteStatement<ShardPreparedStatement, AceObjectPropertiesInt>(ShardPreparedStatement.DeleteAceObjectPropertyInt, prop);
+                        properties.Remove(prop);
+                    }
+                    else
+                    {
+                        // update it
+                        transaction.AddPreparedUpdateStatement<ShardPreparedStatement, AceObjectPropertiesInt>(ShardPreparedStatement.UpdateAceObjectPropertyInt, prop);
+                    }
+                }
+                else if (prop.PropertyValue != null)
+                {
+                    transaction.AddPreparedInsertStatement<ShardPreparedStatement, AceObjectPropertiesInt>(ShardPreparedStatement.InsertAceObjectPropertiesInt, prop);
+                }
+            }
+
```

Comment:
Its one thing to copy/paste a 1 line function (smelly but not too bad), I'm weary of all this copy pasta.